utilities/env_librados.cc

```diff
@@ -0,0 +1,1043 @@
+  // Furthermore, it tries to make this ID at most "max_size" bytes. If such an
+  // ID can be created this function returns the length of the ID and places it
+  // in "id"; otherwise, this function returns 0, in which case "id"
+  // may not have been modified.
+  //
+  // This function guarantees, for IDs from a given environment, two unique ids
+  // cannot be made equal to eachother by adding arbitrary bytes to one of
+  // them. That is, no unique ID is the prefix of another.
+  //
+  // This function guarantees that the returned ID will not be interpretable as
+  // a single varint.
+  //
+  // Note: these IDs are only valid for the duration of the process.
+  size_t GetUniqueId(char* id, size_t max_size) const {
+  #if 0
+    return snprintf(id, max_size, "%016llx",
+             (unsigned long long)h->file->fnode.ino);
+  #endif
+    return 0;
+  };
+
+  //enum AccessPattern { NORMAL, RANDOM, SEQUENTIAL, WILLNEED, DONTNEED };
+
+  void Hint(AccessPattern pattern) {
+  #if 0
+    if (pattern == RANDOM)
+      h->buf.max_prefetch = 4096;
+    else if (pattern == SEQUENTIAL)
+      h->buf.max_prefetch = g_conf->bluefs_max_prefetch;
+  #endif
+  }
+
+  // Remove any kind of caching of data from the offset to offset+length
+  // of this file. If the length is 0, then it refers to the end of file.
+  // If the system is not caching the file contents, then this is a noop.
+  rocksdb::Status InvalidateCache(size_t offset, size_t length) {
+  #if 0
+    fs->invalidate_cache(h->file, offset, length);
+  #endif
+    return rocksdb::Status::OK();
+  }
+};
+
+
+// A file abstraction for sequential writing.  The implementation
+// must provide buffering since callers may append small fragments
+// at a time to the file.
+class LibradosWritableFile : public rocksdb::WritableFile {
+    librados::IoCtx _io_ctx;
+    std::string _fid;
+    const int BUFFER_MAX_SIZE = 1 << 12;
+    librados::bufferlist buffer;
+    int buffer_size;
+ public:
+    LibradosWritableFile(librados::IoCtx io_ctx, std::string fid):
+        _io_ctx(io_ctx), _fid(fid), buffer(), buffer_size(0){
+    }
+    ~LibradosWritableFile() {
+        Sync();
+    }
+
+  // Indicates if the class makes use of unbuffered I/O
+  /*bool UseOSBuffer() const {
+    return true;
+    }*/
+
+  // This is needed when you want to allocate
+  // AlignedBuffer for use with file I/O classes
+  // Used for unbuffered file I/O when UseOSBuffer() returns false
+  /*size_t GetRequiredBufferAlignment() const {
+    return c_DefaultPageSize;
+    }*/
+
+  rocksdb::Status Append(const rocksdb::Slice& data) {
+    LOG_DEBUG("%s|%i\n", _fid.c_str(), (int)data.size());
+    #if 0
+    buffer.append(data.data());
+    buffer_size += data.size();
+
+    if (buffer_size > BUFFER_MAX_SIZE) {
+        return Sync();
+    } else {
+        return rocksdb::Status::OK();
+    }
+    #endif
+    librados::bufferlist buf;
+    buf.append(data.data());
+    int r = _io_ctx.append(_fid, buf, data.size());
```

Comment:
I would do something here like:

1. append to the bufferlist.
2. if bufferlist length > some threshold, call flush()

and then make the Flush() do the _io_ctx write.